File: assets/codebase/rendering/AxisRotationComponent.cs
namespace gameplay.rendering {
    /// <summary>
    /// Rotates the parent entity around one authored local-space axis using frame-rate-independent delta time.
    /// </summary>
    public sealed class AxisRotationComponent : UpdateComponent {
        /// <summary>
        /// Gets or sets the local-space axis used by incremental rotation updates.
        /// </summary>
        public float3 Axis { get; set; }

        /// <summary>
        /// Gets or sets the angular speed in radians per second.
        /// </summary>
        public float AngularSpeedRadiansPerSecond { get; set; }

        /// <summary>
        /// Advances the parent local orientation by one delta-time rotation step around the authored local axis.
        /// </summary>
        public override void Update() {
            base.Update();

            if (Axis == float3.Zero) {
                throw new InvalidOperationException("AxisRotationComponent requires a non-zero axis.");
            }

            float3 normalizedAxis = float3.Normalize(Axis);
            float deltaAngleRadians = AngularSpeedRadiansPerSecond * (float)Core.Instance.FrameDeltaSeconds;
            float4 deltaRotation;
            float4.CreateFromAxisAngle(ref normalizedAxis, deltaAngleRadians, out deltaRotation);

            float4 orientation = Parent.LocalOrientation * deltaRotation;
            orientation.Normalize();
            Parent.LocalOrientation = orientation;
        }
    }
}

File: windows-build-20260529-stack-box-direct/cooked/code/gameplay/AxisRotationComponent.cpp
#ifdef DrawText
#undef DrawText
#endif
#include "AxisRotationComponent.hpp"
#include "UpdateComponent.hpp"
#include "runtime/native_exceptions.hpp"
#include "float3.hpp"
#include "float4.hpp"
#include "Core.hpp"
#include "runtime/native_exceptions.hpp"

AxisRotationComponent::AxisRotationComponent() : AngularSpeedRadiansPerSecond(), Axis()
{
}

float AxisRotationComponent::get_AngularSpeedRadiansPerSecond()
{
return this->AngularSpeedRadiansPerSecond;
}

void AxisRotationComponent::set_AngularSpeedRadiansPerSecond(float value)
{
this->AngularSpeedRadiansPerSecond = value;
}

float3 AxisRotationComponent::get_Axis()
{
return this->Axis;
}

void AxisRotationComponent::set_Axis(float3 value)
{
this->Axis = value;
}

void AxisRotationComponent::Update()
{
UpdateComponent::Update();
    if (this->Axis == float3::get_Zero())
    {
throw new InvalidOperationException("AxisRotationComponent requires a non-zero axis.");
    }
float3 normalizedAxis = float3::Normalize(this->Axis);
const float deltaAngleRadians = this->AngularSpeedRadiansPerSecond * static_cast<float>(Core::get_Instance()->get_FrameDeltaSeconds());
float4 deltaRotation;
float4::CreateFromAxisAngle(normalizedAxis, deltaAngleRadians, deltaRotation);
float4 orientation = this->get_Parent()->get_LocalOrientation() * deltaRotation;
orientation.Normalize();
Parent->set_LocalOrientation(orientation);
}

